# 451 — Locus-Level Methylation–Expression Characterization

## Objective

Characterize locus-level DNA-methylation–gene-expression relationships associated
with the frozen tumor-side methylation context of the three Phase 4 cross-system
consensus transcriptomic programs in the TCGA primary-tumor cohort.

The analysis integrates frozen TCGA DNA-methylation and RNA-seq representations
from Phase 2 with the frozen tumor methylation-component context preserved through
Phase 4.

The objective is to determine whether strongly contributing CpG loci from the
relevant frozen methylation components show reproducible associations with
expression of defensibly annotated genes while preserving lineage structure,
platform boundaries, upstream limitations, and explicit separation between
statistical association and biological mechanism.

Inverse promoter-associated methylation–expression relationships are interpreted
as compatible with an epigenetic regulatory hypothesis. They do not establish
methylation-mediated transcriptional repression or biological causality.

## Frozen upstream inputs

Notebook 451 operates on frozen upstream objects:

- the TCGA primary-tumor multi-omic cohort and exact case-level sample mapping
  from notebook 203;
- the frozen TMM-logCPM RNA-seq representation generated in notebook 205;
- the frozen HM450, HM27, and shared HM27/HM450 DNA-methylation
  representations from notebook 203;
- the frozen methylation probe mapping from notebook 203;
- the frozen methylation ICA probe loadings from notebook 205;
- the three frozen Phase 4 consensus transcriptomic programs and their
  tumor-side methylation-component context; and
- frozen tumor-side confounder covariates from notebook 204.

Upstream cohort membership, sample identity, RNA-seq normalization, methylation
QC, methylation-component identities, component orientations, consensus-program
identities, and program definitions are not recomputed or modified on the basis
of locus-level results.

## Frozen program-context boundary

Notebook 451 is restricted to the tumor-side methylation components already
linked to the three frozen consensus transcriptomic programs:

- `CONSENSUS_TX_01` — `METH_IC128`;
- `CONSENSUS_TX_02` — `METH_IC050`; and
- `CONSENSUS_TX_03` — `METH_IC169` and `METH_IC128`.

Because `METH_IC128` occurs in more than one consensus-program context, statistical
inference is performed on unique CpG–gene units rather than duplicated by program
membership.

Consensus-program and methylation-component memberships are preserved as
downstream contextual annotations.

No locus-level result may rescue, exclude, reweight, rename, reorient, or redefine
a frozen consensus program or its upstream tumor methylation context.

## Locus and annotation boundary

The primary locus universe is defined from the strongest signed loading tails of
the relevant frozen methylation components before methylation–expression
association results are inspected.

For each unique relevant methylation component, the primary locus rule retains:

- the 50 most negative-loading CpGs; and
- the 50 most positive-loading CpGs.

This reuses the pre-existing loading-tail convention established upstream rather
than selecting a new threshold from notebook-451 association results.

CpGs shared across methylation components are retained once as loci while
preserving every applicable component and consensus-program membership.

Biological CpG-to-gene and regulatory-context annotation must come from a
versioned, provenance-tracked, frozen annotation authority before inferential
testing begins.

CpG-to-gene mappings are not selected, rescued, or reassigned according to
observed expression associations, cancer relevance, pathway membership,
druggability, or expected biological interpretation.

## Platform and expression boundary

HM450 is the primary methylation platform for locus-level inference because the
relevant methylation-component discovery was performed in the frozen HM450
analysis space and HM450 provides the larger platform-specific tumor cohort.

HM27 is treated as a prespecified platform-sensitivity layer where locus and
lineage coverage permit evaluation.

HM27 and HM450 samples are not naively pooled into a single primary regression
model.

Gene expression is taken directly from the frozen Phase 2 TMM-logCPM
representation. Notebook 451 does not refit expression filtering or
normalization.

Primary methylation inference uses the M-value transformation convention already
established upstream. Beta values remain available for descriptive interpretation.

## Lineage-aware analytical boundary

Cancer lineage is treated as a required structural component of the analysis.

The primary statistical unit is the unique annotated:

`CpG × gene`

pair.

Primary CpG–gene models are restricted to TCGA projects satisfying prespecified
complete-case and within-project variation requirements.

The primary model is:

`gene_expression ~ CpG_methylation + C(project_id)`

with heteroskedasticity-consistent HC3 uncertainty.

Within-project estimates and leave-one-project-out analyses characterize
directional consistency, lineage dependence, and sensitivity to individual
projects for primary FDR-supported associations.

A pooled association alone is not sufficient to establish cross-project
recurrence.

Cross-project recurrence additionally requires prespecified directional support
across eligible projects and leave-one-project-out stability.

Lineage-aware recurrence describes reproducibility of an association across
eligible TCGA projects. It does not establish biological universality across
cancer types.

## Confounding, multiplicity, and sensitivity analyses

The primary model adjusts for TCGA project while preserving the frozen primary
estimand and maximum eligible HM450 cohort.

Prespecified sensitivity analyses evaluate, where applicable:

- tumor purity;
- proliferation;
- leukocyte fraction for loci belonging to `METH_IC128`; and
- HM27 platform consistency.

The additional cellular-composition sensitivity for `METH_IC128` is motivated by
a frozen upstream interpretive limitation and is not activated according to
notebook-451 results.

Sensitivity analyses characterize stability of primary associations and cannot
rescue associations that fail the primary inferential framework.

Multiplicity is controlled using Benjamini–Hochberg FDR across the complete
frozen unique CpG–gene primary hypothesis family.

FDR families are not subdivided post hoc by consensus program, methylation
component, regulatory annotation, association direction, gene class, or TCGA
project.

## Interpretation boundary

A negative methylation coefficient indicates that higher methylation at the CpG
is associated with lower expression of the annotated gene.

When such a relationship occurs at a promoter-associated locus it may be
described as an inverse promoter-associated methylation–expression relationship
compatible with an epigenetic regulatory hypothesis.

It does not establish:

- promoter silencing;
- methylation-mediated repression;
- a causal regulatory interaction;
- a mechanism of drug resistance;
- a validated biomarker;
- a validated therapeutic target; or
- therapeutic reversibility.

Positive, inverse, heterogeneous, lineage-restricted, technically limited, or
null results are all valid scientific outcomes.

## Scope and methodological boundary

Notebook 451 does not:

- redefine or reoptimize consensus programs;
- refit methylation or transcriptomic latent components;
- assign resistant or sensitive labels to TCGA tumors;
- select loci or genes using pharmacogenomic phenotype;
- infer causal regulation from methylation–expression association;
- choose CpG-to-gene mappings according to association results;
- pool HM27 and HM450 naïvely;
- treat HM27 as independent external biological validation;
- use sensitivity analyses to redefine primary significance;
- promote individual genes as validated targets; or
- use downstream functional, pharmacogenomic, perturbational, or therapeutic
  evidence to select locus-level associations.

The resulting CpG–gene relationships constitute secondary locus-level molecular
context for the frozen consensus programs.

## Expected outputs

Notebook 451 will produce reproducible locus-level characterization under:

`data/processed/secondary_characterization/`

The analysis will preserve:

- the deterministic strong-loading CpG universe and CpG–gene annotation
  eligibility;
- the complete frozen primary CpG–gene association family;
- lineage-aware recurrence characterization;
- prespecified sensitivity evidence;
- promoter/regulatory-context annotation;
- consensus-program and methylation-component context; and
- explicit negative, heterogeneous, and technically limited results.

Only downstream-consumable artifacts justified by the completed analysis will be
persisted and registered. Intermediate diagnostic objects remain reproducible
within the notebook unless separate persistence is required for downstream
provenance or reuse.

In [1]:
# =============================================================================
# Imports
# =============================================================================

import json

import h5py
import numpy as np
import pandas as pd

from pancancer_epigenetics.utils.paths import Paths, project_relative_path

In [2]:
# =============================================================================
# Resolve frozen upstream artifact paths
# =============================================================================

UPSTREAM_ARTIFACT_IDS = (
    "phase2.203.final_sample_mapping",
    "phase2.203.final_hm27_methylation_betas",
    "phase2.203.final_hm450_methylation_betas",
    "phase2.203.final_shared_methylation_betas",
    "phase2.203.final_probe_mapping",
    "phase2.204.confounder_covariates",
    "phase2.205.rna_program_discovery_expression",
    "phase2.205.methylation_candidate_probe_loadings",
    "phase4.401.consensus_transcriptomic_program_catalog",
    "phase4.401.consensus_tumor_arm_context",
)

with Paths.artifact_registry.open("r", encoding="utf-8") as handle:
    artifact_registry = json.load(handle)

upstream_artifact_paths = {
    artifact_id: (
        Paths.root
        / artifact_registry["artifacts"][artifact_id]["path"]
    )
    for artifact_id in UPSTREAM_ARTIFACT_IDS
}

for artifact_id, path in upstream_artifact_paths.items():
    print(f"{artifact_id}: {project_relative_path(path)}")

phase2.203.final_sample_mapping: data/interim/metadata/tcga_primary_tumor_multiomic_final_case_level_sample_mapping.csv
phase2.203.final_hm27_methylation_betas: data/interim/methylation/tcga_primary_tumor_methylation_hm27_final_case_level_beta_values.npy
phase2.203.final_hm450_methylation_betas: data/interim/methylation/tcga_primary_tumor_methylation_hm450_final_case_level_beta_values.npy
phase2.203.final_shared_methylation_betas: data/interim/methylation/tcga_primary_tumor_methylation_shared_hm27_hm450_final_case_level_beta_values.npy
phase2.203.final_probe_mapping: data/interim/metadata/tcga_primary_tumor_methylation_final_probe_mapping.csv
phase2.204.confounder_covariates: data/interim/metadata/tcga_primary_tumor_multiomic_confounder_covariates.csv
phase2.205.rna_program_discovery_expression: data/interim/expression/tcga_primary_tumor_rnaseq_program_discovery_tmm_logcpm.h5
phase2.205.methylation_candidate_probe_loadings: data/processed/tumor_programs/tcga_primary_tumor_hm450_ica_can

In [3]:
# =============================================================================
# Validate frozen upstream input contract
# =============================================================================

upstream_input_contract = []

for artifact_id, path in upstream_artifact_paths.items():
    entry = artifact_registry["artifacts"][artifact_id]

    upstream_input_contract.append(
        {
            "artifact_id": artifact_id,
            "status": entry["status"],
            "exists": path.is_file(),
        }
    )

upstream_input_contract = pd.DataFrame(upstream_input_contract)

print(
    "All inputs frozen:",
    bool(upstream_input_contract["status"].eq("frozen").all()),
)
print(
    "All input files present:",
    bool(upstream_input_contract["exists"].all()),
)

upstream_input_contract

All inputs frozen: True
All input files present: True


,artifact_id,status,exists
0,phase2.203.final_sample_mapping,frozen,True
1,phase2.203.final_hm27_methylation_betas,frozen,True
2,phase2.203.final_hm450_methylation_betas,frozen,True
3,phase2.203.final_shared_methylation_betas,frozen,True
4,phase2.203.final_probe_mapping,frozen,True
5,phase2.204.confounder_covariates,frozen,True
6,phase2.205.rna_program_discovery_expression,frozen,True
7,phase2.205.methylation_candidate_probe_loadings,frozen,True
8,phase4.401.consensus_transcriptomic_program_ca...,frozen,True
9,phase4.401.consensus_tumor_arm_context,frozen,True


In [6]:
# =============================================================================
# Load frozen tabular upstream inputs
# =============================================================================

sample_mapping = pd.read_csv(
    upstream_artifact_paths["phase2.203.final_sample_mapping"]
)

probe_mapping = pd.read_csv(
    upstream_artifact_paths["phase2.203.final_probe_mapping"],
    low_memory=False
)

confounder_covariates = pd.read_csv(
    upstream_artifact_paths["phase2.204.confounder_covariates"]
)

methylation_probe_loadings = pd.read_csv(
    upstream_artifact_paths[
        "phase2.205.methylation_candidate_probe_loadings"
    ]
)

consensus_program_catalog = pd.read_csv(
    upstream_artifact_paths[
        "phase4.401.consensus_transcriptomic_program_catalog"
    ]
)

consensus_tumor_arm_context = pd.read_csv(
    upstream_artifact_paths[
        "phase4.401.consensus_tumor_arm_context"
    ]
)

print("Sample mapping:", sample_mapping.shape)
print("Probe mapping:", probe_mapping.shape)
print("Confounder covariates:", confounder_covariates.shape)
print("Methylation probe loadings:", methylation_probe_loadings.shape)
print("Consensus program catalog:", consensus_program_catalog.shape)
print("Consensus tumor-arm context:", consensus_tumor_arm_context.shape)

Sample mapping: (9965, 30)
Probe mapping: (455355, 14)
Confounder covariates: (9965, 18)
Methylation probe loadings: (10000, 29)
Consensus program catalog: (3, 21)
Consensus tumor-arm context: (4, 25)


In [7]:
# =============================================================================
# Inspect fields required for notebook-451 transformations
# =============================================================================

print(
    "Methylation probe loadings columns:",
    methylation_probe_loadings.columns.tolist(),
)
print()
print(
    "Consensus tumor-arm context columns:",
    consensus_tumor_arm_context.columns.tolist(),
)
print()
print(
    "Sample mapping columns:",
    sample_mapping.columns.tolist(),
)
print()
print(
    "Confounder covariates columns:",
    confounder_covariates.columns.tolist(),
)

Methylation probe loadings columns: ['model_matrix_row_index', 'representation', 'methylation_platform', 'matrix_row_index', 'probe_id', 'source_platform_probe_order', 'source_shared_probe_order', 'probe_missing_beta_fraction', 'probe_qc_eligibility_scope', 'probe_qc_eligible', 'hm27_missing_beta_fraction', 'hm450_missing_beta_fraction', 'hm27_probe_qc_eligible', 'hm450_probe_qc_eligible', 'shared_probe_qc_eligible', 'shared_matrix_row_index', 'median_within_project_variance', 'METH_IC152', 'METH_IC232', 'METH_IC169', 'METH_IC128', 'METH_IC023', 'METH_IC013', 'METH_IC234', 'METH_IC050', 'METH_IC107', 'METH_IC241', 'METH_IC109', 'METH_IC033']

Consensus tumor-arm context columns: ['consensus_program_id', 'candidate_pair', 'rna_component', 'methylation_component', 'structural_family', 'cell_line_program', 'orientation_multiplier', 'final_audit_status', 'scientific_priority', 'strict_hm27_status', 'relationship_form', 'primary_limitation', 'consensus_handling', 'lopo_absolute_shift', 'lop

In [8]:
# =============================================================================
# Define frozen notebook-451 methylation-component context
# =============================================================================

methylation_component_context = (
    consensus_tumor_arm_context[
        [
            "consensus_program_id",
            "candidate_pair",
            "rna_component",
            "methylation_component",
            "final_audit_status",
            "scientific_priority",
            "primary_limitation",
            "consensus_handling",
        ]
    ]
    .drop_duplicates()
    .sort_values(
        ["consensus_program_id", "methylation_component", "candidate_pair"]
    )
    .reset_index(drop=True)
)

relevant_methylation_components = tuple(
    sorted(
        methylation_component_context[
            "methylation_component"
        ].dropna().unique()
    )
)

missing_loading_components = sorted(
    set(relevant_methylation_components)
    - set(methylation_probe_loadings.columns)
)

print(
    "Consensus programs represented:",
    methylation_component_context["consensus_program_id"].nunique(),
)
print(
    "Tumor methylation arms represented:",
    methylation_component_context.shape[0],
)
print(
    "Unique methylation components:",
    relevant_methylation_components,
)
print(
    "All relevant components present in frozen loadings:",
    not missing_loading_components,
)

methylation_component_context

Consensus programs represented: 3
Tumor methylation arms represented: 4
Unique methylation components: ('METH_IC050', 'METH_IC128', 'METH_IC169')
All relevant components present in frozen loadings: True


,consensus_program_id,candidate_pair,rna_component,methylation_component,final_audit_status,scientific_priority,primary_limitation,consensus_handling
0,CONSENSUS_TX_01,CROSS_OMIC_PAIR_04,RNA_IC150,METH_IC128,CONFOUNDED,LOW,Effect fell from original 0.133 to HM450 refit...,Platform support does not rehabilitate biologi...
1,CONSENSUS_TX_02,CROSS_OMIC_PAIR_08,RNA_IC151,METH_IC050,METHOD_LIMITED,MEDIUM,No stable strict HM27 component recovery,Carry as a secondary small-effect candidate; d...
2,CONSENSUS_TX_03,CROSS_OMIC_PAIR_12,RNA_IC184,METH_IC128,CONFOUNDED,LOW_CONDITIONAL,No valid HM27 test and interpretation remains ...,Do not promote automatically; require the exis...
3,CONSENSUS_TX_03,CROSS_OMIC_PAIR_03,RNA_IC184,METH_IC169,METHOD_LIMITED,CONDITIONAL_HIGH,HM450-only matching missed the frozen score-co...,Retain as the main continuous hypothesis but d...


In [10]:
# =============================================================================
# Construct strong-loading CpG membership universe
# =============================================================================

LOADING_TAIL_SIZE = 50

strong_loading_memberships = []

for component in relevant_methylation_components:
    component_loadings = methylation_probe_loadings[
        ["model_matrix_row_index", "probe_id", component]
    ].rename(columns={component: "loading"})

    negative_tail = (
        component_loadings
        .sort_values(
            ["loading", "probe_id"],
            ascending=[True, True],
        )
        .head(LOADING_TAIL_SIZE)
        .assign(
            methylation_component=component,
            loading_tail="negative",
        )
    )

    positive_tail = (
        component_loadings
        .sort_values(
            ["loading", "probe_id"],
            ascending=[False, True],
        )
        .head(LOADING_TAIL_SIZE)
        .assign(
            methylation_component=component,
            loading_tail="positive",
        )
    )

    strong_loading_memberships.extend(
        [negative_tail, positive_tail]
    )

strong_loading_cpg_memberships = (
    pd.concat(
        strong_loading_memberships,
        ignore_index=True,
    )
    [
        [
            "probe_id",
            "model_matrix_row_index",
            "methylation_component",
            "loading",
            "loading_tail",
        ]
    ]
    .sort_values(
        ["methylation_component", "loading_tail", "loading"],
        ascending=[True, True, True],
    )
    .reset_index(drop=True)
)

expected_memberships = (
    len(relevant_methylation_components)
    * 2
    * LOADING_TAIL_SIZE
)

print("Expected component-tail memberships:", expected_memberships)
print(
    "Observed component-tail memberships:",
    strong_loading_cpg_memberships.shape[0],
)
print(
    "Unique strong-loading CpGs:",
    strong_loading_cpg_memberships["probe_id"].nunique(),
)
print(
    "Duplicate membership rows:",
    strong_loading_cpg_memberships.duplicated(
        ["probe_id", "methylation_component", "loading_tail"]
    ).sum(),
)

strong_loading_cpg_memberships.groupby(
    ["methylation_component", "loading_tail"]
).size()

Expected component-tail memberships: 300
Observed component-tail memberships: 300
Unique strong-loading CpGs: 279
Duplicate membership rows: 0


methylation_component  loading_tail
METH_IC050             negative        50
                       positive        50
METH_IC128             negative        50
                       positive        50
METH_IC169             negative        50
                       positive        50
dtype: int64

In [11]:
# =============================================================================
# Construct unique strong-loading CpG locus universe
# =============================================================================

probe_index_counts = (
    strong_loading_cpg_memberships
    .groupby("probe_id")["model_matrix_row_index"]
    .nunique()
)

if not probe_index_counts.eq(1).all():
    raise ValueError(
        "At least one strong-loading CpG maps to multiple model matrix rows."
    )

strong_loading_cpg_universe = (
    strong_loading_cpg_memberships
    .groupby(
        ["probe_id", "model_matrix_row_index"],
        as_index=False,
    )
    .agg(
        n_methylation_components=(
            "methylation_component",
            "nunique",
        ),
        methylation_components=(
            "methylation_component",
            lambda values: "|".join(sorted(set(values))),
        ),
        component_tail_memberships=(
            "methylation_component",
            lambda values: "|".join(
                sorted(set(values))
            ),
        ),
    )
)

component_tail_context = (
    strong_loading_cpg_memberships
    .assign(
        component_tail=lambda frame: (
            frame["methylation_component"]
            + ":"
            + frame["loading_tail"]
        )
    )
    .groupby("probe_id")["component_tail"]
    .agg(lambda values: "|".join(sorted(set(values))))
)

strong_loading_cpg_universe[
    "component_tail_memberships"
] = strong_loading_cpg_universe["probe_id"].map(
    component_tail_context
)

strong_loading_cpg_universe = (
    strong_loading_cpg_universe
    .sort_values("probe_id")
    .reset_index(drop=True)
)

print(
    "Unique strong-loading CpGs:",
    strong_loading_cpg_universe.shape[0],
)
print(
    "CpGs represented in >1 methylation component:",
    int(
        strong_loading_cpg_universe[
            "n_methylation_components"
        ].gt(1).sum()
    ),
)
print(
    "Maximum component memberships for one CpG:",
    int(
        strong_loading_cpg_universe[
            "n_methylation_components"
        ].max()
    ),
)

strong_loading_cpg_universe[
    "n_methylation_components"
].value_counts().sort_index()

Unique strong-loading CpGs: 279
CpGs represented in >1 methylation component: 21
Maximum component memberships for one CpG: 2


n_methylation_components
1    258
2     21
Name: count, dtype: int64

In [12]:
# =============================================================================
# Map strong-loading CpGs to frozen HM450 matrix rows
# =============================================================================

hm450_probe_mapping = (
    probe_mapping.loc[
        probe_mapping["representation"].eq("hm450"),
        [
            "probe_id",
            "matrix_row_index",
            "probe_missing_beta_fraction",
            "probe_qc_eligible",
            "hm450_missing_beta_fraction",
            "hm450_probe_qc_eligible",
        ],
    ]
    .rename(
        columns={
            "matrix_row_index": "hm450_matrix_row_index",
        }
    )
)

hm450_strong_probe_mapping = (
    strong_loading_cpg_universe
    .merge(
        hm450_probe_mapping,
        on="probe_id",
        how="left",
        validate="one_to_one",
    )
    .sort_values("probe_id")
    .reset_index(drop=True)
)

if hm450_strong_probe_mapping["hm450_matrix_row_index"].isna().any():
    raise ValueError(
        "At least one strong-loading CpG lacks an HM450 matrix-row mapping."
    )

print(
    "Strong-loading CpGs:",
    hm450_strong_probe_mapping.shape[0],
)
print(
    "Mapped to HM450 rows:",
    hm450_strong_probe_mapping[
        "hm450_matrix_row_index"
    ].notna().sum(),
)
print(
    "HM450 QC-eligible:",
    hm450_strong_probe_mapping[
        "hm450_probe_qc_eligible"
    ].fillna(False).sum(),
)
print(
    "Unique HM450 matrix rows:",
    hm450_strong_probe_mapping[
        "hm450_matrix_row_index"
    ].nunique(),
)

Strong-loading CpGs: 279
Mapped to HM450 rows: 279
HM450 QC-eligible: 279
Unique HM450 matrix rows: 279


In [13]:
# =============================================================================
# Inspect frozen RNA expression HDF5 structure
# =============================================================================

rna_expression_path = upstream_artifact_paths[
    "phase2.205.rna_program_discovery_expression"
]

with h5py.File(rna_expression_path, "r") as h5_file:
    print("Root keys:", list(h5_file.keys()))
    print("File attributes:", dict(h5_file.attrs))

    rna_logcpm_dataset = h5_file["logcpm"]

    print("logcpm shape:", rna_logcpm_dataset.shape)
    print("logcpm dtype:", rna_logcpm_dataset.dtype)
    print(
        "logcpm attributes:",
        dict(rna_logcpm_dataset.attrs),
    )

Root keys: ['logcpm']
File attributes: {}
logcpm shape: (18123, 9965)
logcpm dtype: float64
logcpm attributes: {'orientation': array([b'genes_x_samples'], dtype='|S16'), 'rhdf5-NA.OK': array([1], dtype=int32)}


In [14]:
# =============================================================================
# Load RNA program-discovery gene metadata
# =============================================================================

RNA_DISCOVERY_GENE_METADATA_PATH = (
    Paths.expression
    / "tcga_primary_tumor_rnaseq_program_discovery_filtered_gene_metadata.csv"
)

if not RNA_DISCOVERY_GENE_METADATA_PATH.is_file():
    raise FileNotFoundError(
        "Frozen RNA program-discovery gene metadata is missing."
    )

rna_discovery_gene_metadata = pd.read_csv(
    RNA_DISCOVERY_GENE_METADATA_PATH
)

with h5py.File(rna_expression_path, "r") as h5_file:
    rna_logcpm_shape = h5_file["logcpm"].shape

if len(rna_discovery_gene_metadata) != rna_logcpm_shape[0]:
    raise ValueError(
        "RNA gene metadata rows do not match the frozen logCPM gene axis."
    )

if len(sample_mapping) != rna_logcpm_shape[1]:
    raise ValueError(
        "Frozen sample mapping rows do not match the logCPM sample axis."
    )

print(
    "RNA gene metadata path:",
    project_relative_path(RNA_DISCOVERY_GENE_METADATA_PATH),
)
print("RNA gene metadata shape:", rna_discovery_gene_metadata.shape)
print("RNA gene metadata columns:", rna_discovery_gene_metadata.columns.tolist())
print(
    "Gene-axis alignment:",
    len(rna_discovery_gene_metadata) == rna_logcpm_shape[0],
)
print(
    "Sample-axis alignment:",
    len(sample_mapping) == rna_logcpm_shape[1],
)

RNA gene metadata path: data/interim/expression/tcga_primary_tumor_rnaseq_program_discovery_filtered_gene_metadata.csv
RNA gene metadata shape: (18123, 7)
RNA gene metadata columns: ['matrix_row_index', 'gene_id', 'gene_name', 'gene_type', 'gene_id_base', 'gene_id_is_versioned', 'expressed_sample_count']
Gene-axis alignment: True
Sample-axis alignment: True


In [15]:
# =============================================================================
# Construct frozen RNA expression gene index
# =============================================================================

rna_expression_gene_index = (
    rna_discovery_gene_metadata
    .assign(
        rna_logcpm_row_index=np.arange(
            len(rna_discovery_gene_metadata),
            dtype=np.int64,
        )
    )
    [
        [
            "rna_logcpm_row_index",
            "matrix_row_index",
            "gene_id",
            "gene_id_base",
            "gene_name",
            "gene_type",
            "gene_id_is_versioned",
            "expressed_sample_count",
        ]
    ]
    .copy()
)

if rna_expression_gene_index["rna_logcpm_row_index"].duplicated().any():
    raise ValueError("RNA logCPM row indices are not unique.")

if rna_expression_gene_index["matrix_row_index"].duplicated().any():
    raise ValueError("Frozen upstream RNA matrix row indices are not unique.")

gene_name_counts = (
    rna_expression_gene_index["gene_name"]
    .dropna()
    .value_counts()
)

gene_id_base_counts = (
    rna_expression_gene_index["gene_id_base"]
    .dropna()
    .value_counts()
)

print("RNA expression genes:", len(rna_expression_gene_index))
print(
    "Unique Ensembl base IDs:",
    rna_expression_gene_index["gene_id_base"].nunique(),
)
print(
    "Ensembl base IDs represented by >1 row:",
    int(gene_id_base_counts.gt(1).sum()),
)
print(
    "Unique gene symbols:",
    rna_expression_gene_index["gene_name"].nunique(),
)
print(
    "Gene symbols represented by >1 row:",
    int(gene_name_counts.gt(1).sum()),
)
print(
    "Missing gene symbols:",
    int(rna_expression_gene_index["gene_name"].isna().sum()),
)

RNA expression genes: 18123
Unique Ensembl base IDs: 18123
Ensembl base IDs represented by >1 row: 0
Unique gene symbols: 18122
Gene symbols represented by >1 row: 1
Missing gene symbols: 0


In [16]:
# =============================================================================
# Characterize RNA gene-symbol ambiguity
# =============================================================================

duplicated_gene_symbols = (
    gene_name_counts.loc[gene_name_counts.gt(1)]
    .index
    .tolist()
)

ambiguous_gene_symbol_rows = (
    rna_expression_gene_index.loc[
        rna_expression_gene_index["gene_name"].isin(
            duplicated_gene_symbols
        )
    ]
    .sort_values(
        ["gene_name", "gene_id_base"]
    )
    .reset_index(drop=True)
)

print(
    "Ambiguous gene symbols:",
    len(duplicated_gene_symbols),
)
print(
    "RNA rows involved:",
    len(ambiguous_gene_symbol_rows),
)

ambiguous_gene_symbol_rows[
    [
        "rna_logcpm_row_index",
        "gene_id",
        "gene_id_base",
        "gene_name",
        "gene_type",
    ]
]

Ambiguous gene symbols: 1
RNA rows involved: 2


,rna_logcpm_row_index,gene_id,gene_id_base,gene_name,gene_type
0,11882,ENSG00000168255.20,ENSG00000168255,POLR2J3,protein_coding
1,18094,ENSG00000285437.1,ENSG00000285437,POLR2J3,protein_coding


In [19]:
# =============================================================================
# Define deterministic RNA gene-identifier lookup tables
# =============================================================================

rna_ensembl_lookup = (
    rna_expression_gene_index[
        [
            "gene_id_base",
            "gene_name",
            "rna_logcpm_row_index",
            "matrix_row_index",
        ]
    ]
    .set_index("gene_id_base")
)

if not rna_ensembl_lookup.index.is_unique:
    raise ValueError(
        "RNA Ensembl base-ID lookup contains duplicated index values."
    )

unambiguous_gene_symbols = gene_name_counts.loc[
    gene_name_counts.eq(1)
].index

rna_symbol_lookup = (
    rna_expression_gene_index.loc[
        rna_expression_gene_index["gene_name"].isin(
            unambiguous_gene_symbols
        ),
        [
            "gene_name",
            "gene_id_base",
            "rna_logcpm_row_index",
            "matrix_row_index",
        ],
    ]
    .set_index("gene_name")
)

if not rna_symbol_lookup.index.is_unique:
    raise ValueError(
        "RNA gene-symbol lookup contains duplicated index values."
    )

print(
    "Ensembl lookup entries:",
    len(rna_ensembl_lookup),
)
print(
    "Unambiguous gene-symbol lookup entries:",
    len(rna_symbol_lookup),
)
print(
    "Excluded ambiguous gene symbols:",
    len(duplicated_gene_symbols),
)
print(
    "Ambiguous symbols:",
    duplicated_gene_symbols,
)

Ensembl lookup entries: 18123
Unambiguous gene-symbol lookup entries: 18121
Excluded ambiguous gene symbols: 1
Ambiguous symbols: ['POLR2J3']


### **CpG biological-annotation checkpoint**

The strong-loading CpG universe and the frozen RNA expression gene axis have now
been defined independently of methylation–expression association results.

Before constructing CpG–gene pairs, notebook 451 requires one authoritative,
versioned biological annotation resource for HM450 probes.

The annotation resource must provide, where available:

- CpG probe identity;
- genomic coordinates;
- genome build;
- gene association;
- promoter or other regulatory-context annotation; and
- sufficient provenance to reproduce the mapping deterministically.

The annotation authority must be frozen before any methylation–expression
association is calculated.

Primary CpG-to-gene assignment will follow the prespecified identifier policy:

1. exact Ensembl base-ID matching when the annotation resource provides a
   compatible stable Ensembl gene identifier;
2. otherwise exact gene-symbol matching only when the symbol is unambiguous in
   the frozen RNA expression matrix.

No fuzzy matching, nearest-gene rescue, manual alias substitution, or
result-driven reassignment is permitted in the primary analysis.

The frozen RNA expression representation contains 18,123 genes with unique
Ensembl base identifiers. Gene-symbol mapping is unambiguous for 18,121 genes.
`POLR2J3` is represented by two distinct Ensembl genes and therefore cannot be
assigned through symbol-only matching.

CpGs mapping to multiple defensible genes will retain every valid CpG–gene
relationship as separate candidate pairs. Repeated transcript-level annotations
of the same CpG–gene relationship will later be collapsed to one unique
inferential CpG–gene unit while preserving annotation metadata.

No methylation–expression inference proceeds until the annotation resource,
version, genome build, provenance, and local frozen representation have been
defined and audited.

### **Frozen CpG annotation authority**

Notebook 451 uses a single prespecified biological annotation authority for
HM450 CpG probes:

- annotation framework: `SeSAMe / sesameData`;
- methylation platform: `HM450`;
- genome build: `hg38`.

This decision is frozen before inspection of any locus-level
methylation–expression association.

The annotation resource will be used only to establish deterministic biological
context for the frozen strong-loading CpG universe, including, where available:

- genomic coordinates;
- CpG-to-gene relationships; and
- promoter or other regulatory-context annotation.

The selected annotation authority does not replace or modify the frozen
technical probe mapping produced in Phase 2. The Phase 2 probe mapping remains
authoritative for matrix-row identity, QC eligibility, platform membership, and
missingness.

Notebook 451 will therefore preserve two distinct annotation layers:

1. **technical probe identity and matrix mapping** inherited from Phase 2; and
2. **biological CpG annotation** derived from the frozen SeSAMe HM450 hg38
   resource.

No alternative annotation source may be substituted after association results
are inspected because it yields more favorable CpG-to-gene mappings or
regulatory interpretation.

The exact SeSAMe / sesameData version, resource identity, retrieval procedure,
local path, and cryptographic identity must be recorded before inferential
analysis begins.

In [20]:
# =============================================================================
# Define frozen CpG annotation resource specification
# =============================================================================

CPG_ANNOTATION_SPEC = {
    "source_database": "Bioconductor",
    "provider": "SeSAMe / sesameData",
    "bioconductor_release": "3.23",
    "package": "sesameData",
    "package_version": "1.30.0",
    "platform": "HM450",
    "genome_build": "hg38",
    "resource_role": "cpg_gene_regulatory_annotation",
}

print("CpG annotation authority:")
for key, value in CPG_ANNOTATION_SPEC.items():
    print(f"  {key}: {value}")

CpG annotation authority:
  source_database: Bioconductor
  provider: SeSAMe / sesameData
  bioconductor_release: 3.23
  package: sesameData
  package_version: 1.30.0
  platform: HM450
  genome_build: hg38
  resource_role: cpg_gene_regulatory_annotation


In [21]:
# =============================================================================
# Inspect raw-data registry for CpG annotation authority
# =============================================================================

with Paths.raw_data_registry.open("r", encoding="utf-8") as handle:
    raw_data_registry = json.load(handle)

registry_text = json.dumps(
    raw_data_registry,
    sort_keys=True,
).lower()

annotation_registry_terms = (
    "sesamedata",
    "sesame / sesamedata",
)

existing_annotation_registration = any(
    term in registry_text
    for term in annotation_registry_terms
)

print(
    "Existing SeSAMe / sesameData registry entry:",
    existing_annotation_registration,
)

Existing SeSAMe / sesameData registry entry: False


In [22]:
# =============================================================================
# Define planned sesameData HM450 annotation registry entry
# =============================================================================

SESAME_ANNOTATION_DIR = (
    Paths.tcga
    / "annotations"
    / "sesame_hm450_hg38"
)

SESAME_ANNOTATION_FILENAME = (
    "sesame_hm450_hg38_probe_annotation.csv"
)

SESAME_ANNOTATION_PATH = (
    SESAME_ANNOTATION_DIR
    / SESAME_ANNOTATION_FILENAME
)

sesame_annotation_registry_entry = {
    "provenance_mode": "file_managed",
    "source_database": "Bioconductor",
    "provider": "SeSAMe / sesameData",
    "release": CPG_ANNOTATION_SPEC["bioconductor_release"],
    "package": CPG_ANNOTATION_SPEC["package"],
    "package_version": CPG_ANNOTATION_SPEC["package_version"],
    "platform": CPG_ANNOTATION_SPEC["platform"],
    "genome_build": CPG_ANNOTATION_SPEC["genome_build"],
    "status": "planned",
    "canonical_dir": project_relative_path(
        SESAME_ANNOTATION_DIR
    ),
    "files": {
        SESAME_ANNOTATION_FILENAME: {
            "role": "cpg_gene_regulatory_annotation",
            "status": "planned",
        }
    },
}

print(
    "Planned annotation path:",
    project_relative_path(SESAME_ANNOTATION_PATH),
)
print()
print(
    json.dumps(
        sesame_annotation_registry_entry,
        indent=2,
    )
)

Planned annotation path: data/raw/tcga/annotations/sesame_hm450_hg38/sesame_hm450_hg38_probe_annotation.csv

{
  "provenance_mode": "file_managed",
  "source_database": "Bioconductor",
  "provider": "SeSAMe / sesameData",
  "release": "3.23",
  "package": "sesameData",
  "package_version": "1.30.0",
  "platform": "HM450",
  "genome_build": "hg38",
  "status": "planned",
  "canonical_dir": "data/raw/tcga/annotations/sesame_hm450_hg38",
  "files": {
    "sesame_hm450_hg38_probe_annotation.csv": {
      "role": "cpg_gene_regulatory_annotation",
      "status": "planned"
    }
  }
}
